perf/reader.go

```diff
@@ -15,6 +15,15 @@ import (
 	"golang.org/x/xerrors"
 )
 
+const (
+	// `perf_raw_record` contains a trailing `u32 size`, whose length is
+	// included in `perf_event_header.size`, the size of the record's buffer.
+	// When reading raw perf samples from the ring, this many bytes need to be
+	// popped from the end of the data to avoid returning garbage once the ring
+	// has wrapped around.
+	perfRawRecordTailLength = 4
```

Comment:
readRawSample uses the value of `u32 size`, not `perf_event_header.size`. From `man perf_event_open`:

```
size, data[size]
                      If PERF_SAMPLE_RAW is enabled, then a 32-bit value indicating size is included
                      followed by an array of 8-bit values of length size.  The  values  are  padded
                      with 0 to have 64-bit alignment.
```

What size is the sample you are submitting? Are you sure this isn't padding for the purpose of alignment?

Maybe you can add a test case that demonstrates the issue.

Reply:
Indeed, now that part of the documentation makes more sense to me, but it's still not quite correct. I've taken another look at [`perf_output_sample()`](https://elixir.bootlin.com/linux/v5.5.13/source/kernel/events/core.c#L6611) and [`perf_prepare_sample()`](https://elixir.bootlin.com/linux/v5.5.13/source/kernel/events/core.c#L6827). Still not sure if the buffer sizing logic (in the kernel) is correct. The change I suggested here indeed breaks on payloads of 4 bytes and smaller; where popping a flat 4 bytes off the end of the array will just yield an empty slice.

The samples I'm submitting are over 100 bytes, so this didn't surface in my testing. Here, the last 4 bytes of the payload were zeroed until the 40-50th event, when they would start containing garbage, likely when the ring has  wrapped at least once.
```
 	want := []byte{1, 2, 3, 4, 4, 0, 0, 0, 0, 0, 0, 0}
                                               ^ these  ^
```

I've analyzed the [segment in `perf_prepare_sample()`](https://elixir.bootlin.com/linux/v5.5.13/source/kernel/events/core.c#L6827) and run the calculations by hand.

When sending a sample with `mustOutputSamplesProg(t, 4)`:
```
fragsize = 4

sum = 4
size = 8 (4 + sizeof32 = 8) aligned to sizeof64

raw->size = 4 (size - sizeof32)
frag->pad = 4 (raw->size - sum)

header->size += 8 (size)
```
Test output: `TestPerfReader: reader_test.go:62: [1 2 3 4]`, 4 bytes.

With `mustOutputSamplesProg(t, 5)`:
```
fragsize = 5

sum = 5
size = 16   (5 + sizeof32 = 9) aligned to sizeof64

raw->size = 12 (size - sizeof(32))
frag->pad = 7 (raw->size - sum)

header->size += 16 (size)
```
Test output: `TestPerfReader: reader_test.go:62: [1 2 3 4 4 0 0 0 0 0 0 0]`, 12 bytes.

In `perf_output_sample()`: `perf_output_put(handle, raw->size);`

The length of the **buffer** read from the ring seems to be of length `raw->size`, but its contents are of length `raw->size - sizeof(u32)`, with a floor of 4 bytes.

Does clamping the sample size to a minimum of 4 bytes sound reasonable @lmb ?

Reply:
I don't think that works, because there is no way to recover the value of `pad` from the information stored in the ring buffer. Any sample between 5 and 5+7 bytes will have the same length in the buffer, but we can't tell where it ends. This is a shortcoming in the perf ring API.

The way you solve this is to include a length field in your raw sample if it is variable length. If it's fixed length you just ignore the tail.

Not sure where the garbage you mention comes from. I kind of assumed that the perf buffer zeroes out pad bytes.

Reply:
Hmm, from reading that `man perf_event_open` snippet, I'm under the impression that the perf ring API is supposed to deliver buffers to userspace that are 64-bit aligned, except the alignment is incorrect. These are opaque buffers that the API itself knows nothing about. If the user wants to use those buffers to encode payloads with type-length-value and ship them to userspace, that's perfectly possible, but at least the buffers themselves should be 64-bit aligned.  That's not happening correctly at the moment, I would expect both `fragsize` 4 and 5 to yield 8 bytes. `pad` should just be used by the kernel internally to determine how many zeroes to append to the buffer so the alignment is correct; not sure why userspace would need access to that.

The initial idea of popping those 4 bytes came from here: https://github.com/iovisor/gobpf/blob/25bfcb50b85f18195f09cf4e6d9d525e985b4889/elf/perf.go#L208. I noticed the difference while migrating from gobpf, because there, whatever I wrote to the ring within BPF came out (64-bit aligned) in userspace, as expected. Same in BCC (Python).

I'll see if I can reproduce the garbage reads.

Reply:
But then the unconditional size - 4 doesn't make sense either, no? If you submit 10 bytes it would cut off the last 2 bytes from the sample.

Reply:
> But then the unconditional size - 4 doesn't make sense either, no?

Oh, of course. :) You are 100% right. That assumption only holds when sending structs to userspace, which 99% (if not more) of tracepoint/kprobe examples in bcc are doing. Structs will always be binpacked and aligned to 64 bits, which will always yield 4 extra bytes due to that kernel bug.

Looks like this cannot/should not be addressed in this library, but IMO it should be documented on the API, as promising 64-bit alignment is misleading. It ends up being more like '64-bit aligned with a possible extra 32 bits of uninitialized memory'.

Sorry for the annoying format, but I have a repro that proves reads return garbage after wrapping around the ring's page boundary for the first time. Add this to `reader_test.go`:

```go
func TestPerfReaderGarbage(t *testing.T) {
	sampleLen := 64

	prog, events := mustOutputSamplesProg(t, sampleLen)
	defer prog.Close()
	defer events.Close()

	rd, err := NewReader(events, 4096)
	if err != nil {
		t.Fatal(err)
	}
	defer rd.Close()

	b := 0
	for i := 0; i < 100; i++ {
		if _, _, err := prog.Test(make([]byte, 14)); err != nil {
			t.Fatal(err)
		}

		record, err := rd.Read()
		if err != nil {
			t.Fatal("Can't read samples:", err)
		}

		// Approximate the total amount of bytes read from the ring.
		// 4 bytes added to account for padding alignment between the
		// end of the sample (68 bytes) and the next header.
		b += binary.Size(perfEventHeader{}) + len(record.RawSample) + 4

		// The sampleLen'th byte should be padding.
		fmt.Println("iter", i, "after", b, "bytes:", record.RawSample)
		if record.RawSample[sampleLen] != 0 {
			t.Fail()
		}
	}
}
```

Returns the following output:

```
iter 49 after 4000 bytes: ...  4 3 2 1 0 0 0 0]
iter 50 after 4080 bytes: ...  4 3 2 1 0 0 0 0]
iter 51 after 4160 bytes: ...  4 3 2 1 1 2 3 4]
iter 52 after 4240 bytes: ...  4 3 2 1 1 2 3 4]
```

Thanks a lot for the reviews! Let me know what you think wrt. adding this as documentation to the API.

Reply:
Interesting, is that kernel behaviour or a bug in the library?

Re documentation, I'm all for it. I was trying to do that with https://github.com/cilium/ebpf/blob/master/perf/reader.go#L62-L64 but reading that again it's not really very useful. It should mention that there are between 1-7 bytes of trailing garbage, and that you need your own size if you have variable length. Maybe you can add an example to the package?

Reply:
The library just calls `io.ReadFull()` on buffers obtained from epoll, so there's not much that can go wrong there. I think in `perf_prepare_sample()`, instead of:

```
			size = round_up(sum + sizeof(u32), sizeof(u64));
			raw->size = size - sizeof(u32);
			frag->pad = raw->size - sum;
```

it should be something along the lines of:

```
			size = round_up(sum + sizeof(u32), sizeof(u64));
			raw->size = round_up(sum, sizeof(64));
			frag->pad = raw->size - sum;
```

I'll play around with it a bit. Looks like there's been a similar observation on the bcc bug tracker: https://github.com/iovisor/bcc/issues/2432.

Reply:
I'm going to stop chasing this; to me it's clear that the buffer size calculation is not working as documented. There are probably multiple generations of perf tools that already depend on this behaviour, so not sure if this makes sense to fix upstream. I played around with a couple of builds, and this is what ends up correctly aligning the header and the `data` buffer: 

```c
void perf_prepare_sample(
...
			/* align the size of the data portion first */
			raw->size = round_up(sum, sizeof(u64));
			/* amount of padding required at end of data portion */
			frag->pad = raw->size - sum;
			/* calculate size of the sample on the ring buffer:
			   | size u32 | + | data[] aligned to u64 | */
			size = raw->size + sizeof(u32);
```

Lastly, the reason for the garbage reads from the ring is because the padding is never actually zeroed.. `__output_skip` is a no-op function.

```c
void perf_output_sample(
...
			if (frag->pad)
				__output_skip(handle, NULL, frag->pad);
```

In summary, I'd like to modify the documentation of `Record.RawSample` to reflect that it will:
- have a length of `(length of submitted sample + 32 bits), aligned to 64 bits, minus 32 bits`, and
- contain between 0 and 7 bytes of trailing garbage from the ring depending on the input sample's length.

Does this look correct? We should probably also add a hint to `Reader.Read()` about this.

A few examples:
- in: 4 bytes, out: 4 bytes
- in: 5 bytes, out: 12 bytes, 7 uninitialized
- in: 12 bytes, out: 12 bytes
- in: 13 bytes, out: 20 bytes, 7 uninitialized
- etc.

Reply:
> I'm going to stop chasing this; to me it's clear that the buffer size calculation is not working as documented. There are probably multiple generations of perf tools that already depend on this behaviour, so not sure if this makes sense to fix upstream.

I think that's sensible, yes.

> Lastly, the reason for the garbage reads from the ring is because the padding is never actually zeroed.. `__output_skip` is a no-op function.
> 
> ```c
> void perf_output_sample(
> ...
> 			if (frag->pad)
> 				__output_skip(handle, NULL, frag->pad);
> ```

This might be fix-worthy upstream.

> In summary, I'd like to modify the documentation of `Record.RawSample` to reflect that it will:
> 
>     * have a length of `(length of submitted sample + 32 bits), aligned to 64 bits, minus 32 bits`, and
> 
>     * contain between 0 and 7 bytes of trailing garbage from the ring depending on the input sample's length.

I think you can drop the first bullet point, seems like a perf buffer implementation detail to me. The second point explains the issue more concisely.

> Does this look correct? We should probably also add a hint to `Reader.Read()` about this.

SGTM.

Reply:
> This might be fix-worthy upstream.

Good idea, submitted this: https://lore.kernel.org/patchwork/patch/1244339.

> I think you can drop the first bullet point, seems like a perf buffer implementation detail to me. The second point explains the issue more concisely.

Agreed, will address all feedback soon. Thanks again for the reviews!